src/notation/internal/engravingfontscontroller.cpp

```diff
@@ -0,0 +1,130 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "engravingfontscontroller.h"
+
+#include "engraving/infrastructure/smufl.h"
+#include "engraving/internal/engravingfont.h"
+#include "draw/internal/ifontsdatabase.h"
+#include "log.h"
+
+#include <QDirIterator>
+#include <QStandardPaths>
+
+using namespace mu::notation;
+
+std::string EngravingFontsController::moduleName() const
+{
+    return "engraving_fonts";
+}
+
+void EngravingFontsController::init()
+{
+    // Standard locations as described in https://w3c.github.io/smufl/latest/specification/font-metadata-locations.html
+
+    // These standard location roughly match up with what the following returns, but some adjustments are needed.
+    QStringList systemFontsPaths = QStandardPaths::standardLocations(QStandardPaths::GenericDataLocation).first(2);
+
+#if defined(Q_OS_WIN)
+    // On Windows, the second standard location returned by Qt is %ProgramData%, but we want %CommonProgramFiles%
+    systemFontsPaths[1] = qgetenv("CommonProgramFiles").replace("\\", "/");
+#elif defined(Q_OS_LINUX)
+    // On Unix systems, we want $XDG_DATA_DIRS and $XDG_DATA_HOME
+    QStringList xdgDataDirs = QString::fromLocal8Bit(qgetenv("XDG_DATA_DIRS")).split(':');
+    systemFontsPaths = xdgDataDirs << qgetenv("XDG_DATA_HOME");
+#endif
+
+    // The first location is the system-wide location, so we should iterate in reverse order so that
+    // user fonts take precedence over system fonts
+    for (QString path : systemFontsPaths) {
+        scanDirectory(path + "/SMuFL/Fonts", false);
+    }
+
+    // Additionally, also support a MuseScore-specific location
+    auto musicFontsPath = configuration()->userMusicFontPathChanged();
+    musicFontsPath.onReceive(this, [this](const muse::io::path_t& dir) {
+        scanDirectory(dir, true);
+    });
+
+    scanDirectory(configuration()->userMusicFontPath(), true);
+}
+
+void EngravingFontsController::scanDirectory(const muse::io::path_t& path, bool isPrivate) const
+{
+    using namespace muse::draw;
+
+    std::shared_ptr<IFontsDatabase> fdb = ioc()->resolve<IFontsDatabase>(moduleName());
+    engravingFonts()->clearUserFonts();
+
+    QDirIterator iterator(path.toQString(), QDir::Dirs | QDir::NoDotAndDotDot | QDir::Readable);
+
+    while (iterator.hasNext()) {
+        QString fontDir = iterator.next();
+        muse::String fontName;
+
+        muse::io::path_t symbolFontPath;
+        muse::io::path_t textFontPath;
+        QDirIterator fontFiles(iterator.filePath(), { "*.otf", "*.ttf" }, QDir::Files);
```

Comment:
I'm afraid it's going to be a bit more complicated:

Currently, you're looking for font files in (subdirs of) the given `path`, which works well for fonts from `configuration()->userMusicFontPath()` (because the design with that is that the user puts both the otf/ttf and the metadata file side-by-side in a subdirectory there).

But for fonts from `systemFontsPaths`, only the metadata file is stored in that path; the otf/ttf is installed in the default system directory for fonts. 

Scanning all system fonts directories for fonts, and then looking for corresponding metadata files, seems not very efficient. The other way around works potentially better, i.e. scan for metadata files and then check if there is a corresponding font in the font database (note that in this case, you probably don't have to load the font by reading the otf/ttf, because it is already in the font database). 

So you'll probably end up with two scanning functions: one for fonts from `configuration()->userMusicFontPath()`, and one for "installed" fonts, in either the global or user-local system font directories.

Reply:
This is the last thing I need to get working, and it's proving to be quite an annoyance...

It seems that you [can't get the file path to a system font via QFont](https://www.qtcentre.org/threads/56880-How-to-find-the-filepath-to-a-font), which is needed to add a font to `IFontsDatabase`... any ideas on how to deal with this?

Reply:
Well, I'm just manually searching through `QStandardPaths::standardLocations(QStandardPaths::FontsLocation)` for now. (With an exception for Windows, since it doesn't include `%LOCALAPPDATA%/Microsoft/Windows/Fonts` 🙃 )